### PR TITLE
Bump dependency version for k8s.io client-go and apimachinery to kubernetes-1.16.2

### PR DESCRIPTION
**Pull request by maxbischoff**
This client-go library uses an older dependency on k8s.io/client-go and k8s.io/apimachinery, which are incompatible with current versions of [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) and [controller tools e.g. controller-gen](https://github.com/kubernetes-sigs/controller-tools).
This PR updates the dependencies to the latest stable kubernetes release. This requires an update of [the istio-testing/build-tools docker image](https://github.com/istio/tools) as well (PR will follow shortly). As soon as that PR is merged, the new image version will be set in the Makefile. Until then you can verify my changes with `IMG=nyli/build-tools:kubernetes-1.16.2 make` (build-tools image generated from https://github.com/maxbischoff/tools/commit/4b686b1d0a8fad1cb76f7b0ebdc589b35838705f).

### CODE REVIEW

**Comment by maxbischoff** [comment]
If `kubernetes-1.16.2` is to recent for you, I would also offer to create a PR for build-tools and this repo with `kubernetes-1.15.5` dependencies.

**Comment by maxbischoff** [comment]
I also opened a PR for setting the image for prow, as gencheck currently cannot work. I am unsure if this is the correct way to do this. If you could provide some feedback that would be great

**Comment by jasonwzm** [comment]
@askmeegs does this relate to the version discrepancy issues you were seeing?

File: go.mod [hunk 1]
```diff
@@ -25,9 +20,6 @@ require (
 	gopkg.in/check.v1 v1.0.0-20190902080502-41f04d3bba15 // indirect
 	gopkg.in/inf.v0 v0.9.1 // indirect
 	istio.io/api v0.0.0-20191015210738-bfa91e88abf1
-	k8s.io/api v0.0.0-20190817221950-ebce17126a01 // indirect
-	k8s.io/apimachinery v0.0.0-20190817221809-bf4de9df677c
-	k8s.io/client-go v10.0.0+incompatible
-	k8s.io/klog v0.4.0 // indirect
-	k8s.io/kube-openapi v0.0.0-20190816220812-743ec37842bf // indirect
+	k8s.io/apimachinery v0.0.0-20191004115801-a2eda9f80ab8
+	k8s.io/client-go v0.0.0-20191016111102-bec269661e48
```
Comment (hzxuzhonghu) [inline go.mod]: Do we need so newer dependency, whatever it should keep consistent with istio repo. Which is still 1.15

Comment (ayj) [reply go.mod]: See https://github.com/istio/istio/blob/156d828d379be772f8d04cac342210e2f04b97b2/go.mod#L168-L219. Looks like we need some extra care to keep k8s deps pinned to a specific version.

**Comment by geeknoid** [comment]
Please use 'make update-common' to update the top-level Makefile instead of manually copying it. Doing so will update the the .commonfiles.sha file in particular.

File: go.mod (hunk 1 above)
Comment (hzxuzhonghu) [reply go.mod]: IC

Comment (hzxuzhonghu) [reply go.mod]: But @howardjohn used `kubernetes-1.16.1` i think

Comment (hzxuzhonghu) [reply go.mod]: OK, Should be ok

**Comment by howardjohn** [comment]
the version here actually doesn't need to match istio/istio at all, right?
Since the repo is independent?

**Comment by maxbischoff** [comment]
@geeknoid using `make update-common` results in build tools being set to `build-tools:2019-10-11T13-37-52`. Do I have to change that version somewhere else?

**Comment by geeknoid** [comment]
make update-common gets you the latest set of files from the istio/common-files repo. In that repo, the file `files/Makefile` defines the IMG variable which is the current container to use. So that's what you're getting. 

I've got a PR waiting to update this (https://github.com/istio/common-files/pull/109) but it's waiting for @sdake 's approval.

**Comment by maxbischoff** [comment]
The updated image from the commons Makefile cannot be used for building this, it seems like the old generators are in that image. With `gcr.io/istio-testing/build-tools:2019-10-22T04-14-16` the build works. Do you have an idea why?

failed build:

```bash
make
Building with the build container: gcr.io/istio-testing/build-tools:2019-10-18T14-38-25.
# generate kube api type wrappers for istio types
# generate deepcopy for kube api types
# generate clientsets for kube api types
# generate listers for kube api types
# generate informers for kube api types
# verifying k8s client
# istio.io/client-go/pkg/clientset/versioned/typed/authentication/v1alpha1
pkg/clientset/versioned/typed/authentication/v1alpha1/authentication_client.gen.go:63:32: undefined: serializer.DirectCodecFactory
# istio.io/client-go/pkg/clientset/versioned/typed/config/v1alpha2
pkg/clientset/versioned/typed/config/v1alpha2/config_client.gen.go:93:32: undefined: serializer.DirectCodecFactory
# istio.io/client-go/pkg/clientset/versioned/typed/rbac/v1alpha1
pkg/clientset/versioned/typed/rbac/v1alpha1/rbac_client.gen.go:73:32: undefined: serializer.DirectCodecFactory
# istio.io/client-go/pkg/clientset/versioned/typed/networking/v1alpha3
pkg/clientset/versioned/typed/networking/v1alpha3/networking_client.gen.go:83:32: undefined: serializer.DirectCodecFactory
# istio.io/client-go/pkg/clientset/versioned/typed/security/v1beta1
pkg/clientset/versioned/typed/security/v1beta1/security_client.gen.go:58:32: undefined: serializer.DirectCodecFactory
Makefile.core.mk:112: recipe for target 'build-k8s-client' failed
make: *** [build-k8s-client] Error 2
Makefile:92: recipe for target 'default' failed
make: *** [default] Error 2
```

successful build with newer image:

```bash
IMG=gcr.io/istio-testing/build-tools:2019-10-22T04-14-16 make
Building with the build container: gcr.io/istio-testing/build-tools:2019-10-22T04-14-16.
# generate kube api type wrappers for istio types
# generate deepcopy for kube api types
# generate clientsets for kube api types
# generate listers for kube api types
# generate informers for kube api types
# verifying k8s client
```

**Comment by maxbischoff** [comment]
The version `gcr.io/istio-testing/build-tools:2019-10-18T14-38-25` from https://github.com/istio/common-files/pull/109 was built for the release-1.4 branch of istio tools: https://prow.istio.io/view/gcs/istio-prow/logs/tools-containers_release-1.4/2 

`gcr.io/istio-testing/build-tools:2019-10-18T14-31-21` is the image from around the same time, built in https://prow.istio.io/view/gcs/istio-prow/logs/tools-containers/76 for master. For that image the build works:

```bash
IMG=gcr.io/istio-testing/build-tools:2019-10-18T14-31-21 make
Building with the build container: gcr.io/istio-testing/build-tools:2019-10-18T14-31-21.
# generate kube api type wrappers for istio types
# generate deepcopy for kube api types
# generate clientsets for kube api types
# generate listers for kube api types
# generate informers for kube api types
# verifying k8s client
```

**Comment by geeknoid** [comment]
Oh, good catch. For now, you can update the IMG value in the top-level
Makefile to an image that works and check that in. I'll make sure the value
in common-files is updated as well. I need to make changes in that repo
today for other reasons.

**Comment by maxbischoff** [comment]
I find it sligthly confusing, that `make gen-check` requires `make tidy` to have been run but only states that `make` should be run to fix the errors. Should `tidy` be a dependency of the main rule?

**Comment by maxbischoff** [comment]
@geeknoid can we get this merged?

**Comment by geeknoid** [comment]
I'll fix the error message that gets printed to tell you to run the right things.

**Pull request merged**